Source/warningreporting/WarningReportingUnit.cpp

```diff
@@ -113,112 +106,72 @@ namespace WarningReporting {
-        #endif
+#ifdef __WINDOWS__
+        string doorBell("127.0.0.1:62002");
+#else
+        string doorBell(Core::Directory::Normalize(pathName) + CyclicBufferName + ".doorbell");
+#endif
 
         Core::SystemInfo::SetEnvironment(WARNINGREPORTING_CYCLIC_BUFFER_FILENAME, fileName);
         Core::SystemInfo::SetEnvironment(WARNINGREPORTING_CYCLIC_BUFFER_DOORBELL, doorBell);
 
-        return (Open(doorBell, fileName));
+        return Open(doorBell, fileName);
     }
 
     uint32_t WarningReportingUnit::Close()
     {
-        m_Admin.Lock();
-
-        //ASSERT(m_OutputChannel != nullptr);
-
-        if (m_OutputChannel != nullptr) {
-            delete m_OutputChannel;
-        }
-
-        m_OutputChannel = nullptr;
-
-        m_Admin.Unlock();
-
-        return (Core::ERROR_NONE);
-    }
-
-    void WarningReportingUnit::Announce(IWarningReportingUnit::IWarningReportingControl& Category)
-    {
-        m_Admin.Lock();
-
-        m_Categories.push_back(&Category);
+        Core::SafeSyncType<Core::CriticalSection> guard(_adminLock);
+        
+        _outputChannel.reset(nullptr);
 
-        m_Admin.Unlock();
+        return Core::ERROR_NONE;
     }
 
-    void WarningReportingUnit::Revoke(IWarningReportingUnit::IWarningReportingControl& Category)
+    void WarningReportingUnit::Announce(IWarningReportingUnit::IWarningReportingControl& category)
     {
-        m_Admin.Lock();
-
-        std::list<IWarningReportingUnit::IWarningReportingControl*>::iterator index(std::find(m_Categories.begin(), m_Categories.end(), &Category));
-
-        if (index != m_Categories.end()) {
-            m_Categories.erase(index);
-        }
-
-        m_Admin.Unlock();
+        Core::SafeSyncType<Core::CriticalSection> guard(_adminLock);
+    
+        _categories[category.Category()] = &category;
     }
 
-    WarningReportingUnit::Iterator WarningReportingUnit::GetCategories()
+    void WarningReportingUnit::Revoke(IWarningReportingUnit::IWarningReportingControl& category)
     {
-        return (Iterator(m_Categories));
+        Core::SafeSyncType<Core::CriticalSection> guard(_adminLock);
+        
+        _categories.erase(category.Category());
     }
 
-    uint32_t WarningReportingUnit::SetCategories(const bool enable, const char* category)
+    std::list<string> WarningReportingUnit::GetCategories()
     {
-        uint32_t modifications = 0;
-
-        ControlList::iterator index(m_Categories.begin());
-
-        while (index != m_Categories.end()) {
-            const char* thisCategory = (*index)->Category();
-
-            if (category != nullptr) {
-                if ( ::strcmp(category, thisCategory) == 0 ) {
-                    modifications++;
-                    (*index)->Enabled(enable);
-                }
-            } else {
-                //Disable/Enable traces for all modules
-                modifications++;
-                (*index)->Enabled(enable);
-            }
-
-            index++;
+        std::list<string> result;
```

Comment:
Didn't check, perhaps this is not called while mutations are possible but just to make sure, shouldn't this be protected by  _adminlock?